docs/src/main/asciidoc/extension-authors-guide.adoc

```diff
@@ -565,3 +568,483 @@ memory in a runtime class loader. To have these classes written out to disk for
 ----
 ./mvnw clean install -Dquarkus.debug.generated-classes-dir=./target/app-generated-classes
 ----
+
+
+== Sample Test Extension
+We have an extension that is used to test for regressions in the extension processing. It is located in {quarkus-tree-url}/core/test-extension directory. In this section we touch on some of the tasks an extension
+author will typically need to perform using the test-extension code to illustrate how the task could be done.
+
+=== Features and Capabilities
+When you start a {project-name} instance, you will see a line like the one highlighted in this example:
+
+.Example Startup Lines
+[source, bash]
+----
+2019-03-22 14:02:37,884 INFO  [io.quarkus] (main) Quarkus 999-SNAPSHOT started in 0.061s. 
+2019-03-22 14:02:37,884 INFO  [io.quarkus] (main) Installed features: [cdi, test-extension] <1>
+----
+<1> A list of features installed in the runtime image
+
+The features listed reflect the types of extensions that are installed. An extension declares its features using a <<Build Step Processors>> method that produces a `FeatureBuildItem` like this TestProcessor#featureBuildItem() method example:
+
+
+.TestProcessor#featureBuildItem()
+[source,java]
+----
+    /**
+     * Register a extension capability and feature
+     * 
+     * @return test-extension feature build item
+     */
+    @BuildStep(providesCapabilities = "io.quarkus.test-extension")
+    FeatureBuildItem featureBuildItem() {
+        return new FeatureBuildItem("test-extension");
+    }
+----
+
+The feature name can be used to select the extension when creating a project using the {project-name} maven plugin as shown in this example taken from the Writing JSON REST Services guide where the "resteasy-jsonb" feature is referenced:
+
+[source,shell,subs=attributes+]
+----
+mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create \
+    -DprojectGroupId=org.acme \
+    -DprojectArtifactId=rest-json \
+    -DclassName="org.acme.rest.json.FruitResource" \
+    -Dpath="/fruits" \
+    -Dextensions="resteasy-jsonb"
+----
+
+=== Bean Defining Annotations
+The CDI layer processes CDI beans that are either explicitly registered or that it discovers based on bean defining annotations as defined in http://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#bean_defining_annotations">[2.5.1. Bean defining annotations]. You can expand this set of annotations to include annotations your extension processes using a `BeanDefiningAnnotationBuildItem` as shown in this `TestProcessor#registerBeanDefinningAnnotations` example:
+
+.Register a Bean Definining Annotation
+[source,java]
+----
+import javax.enterprise.context.ApplicationScoped;
+import org.jboss.jandex.DotName;
+import io.quarkus.extest.runtime.TestAnnotation;
+
+public final class TestProcessor {
+    static DotName TEST_ANNOTATION = DotName.createSimple(TestAnnotation.class.getName());
+    static DotName TEST_ANNOTATION_SCOPE = DotName.createSimple(ApplicationScoped.class.getName());
```

Comment:
You can replace this with `io.quarkus.arc.processor.BuiltinScope.APPLICATION.getInfo().getDotName()`. Maybe, we should add a convenient method like `BuiltinScope.getName()`. Also the defaultScope is not mandatory - `@Dependent` is used by default.